File: 04_BehaviouralDesignPatterns/07_VisitorPattern/Main.java

import java.util.ArrayList;
import java.util.List;

interface Item {
    void accept(ItemVisitor visitor);
}

interface ItemVisitor {
    void visit(PhysicalProduct physicalProduct);

    void visit(DigitalProduct digitalProduct);

    void visit(GiftCard giftCard);
}

class PhysicalProduct implements Item {
    String name;
    double weight;

    public PhysicalProduct(String name, double weight) {
        this.name = name;
        this.weight = weight;
    }

    @Override
    public void accept(ItemVisitor visitor) {
        visitor.visit(this);
    }
}

class DigitalProduct implements Item {
    String name;
    double size;

    public DigitalProduct(String name, double size) {
        this.name = name;
        this.size = size;
    }

    @Override
    public void accept(ItemVisitor visitor) {
        visitor.visit(this);
    }
}

class GiftCard implements Item {
    String code;
    double value;

    public GiftCard(String code, double value) {
        this.code = code;
        this.value = value;
    }

    @Override
    public void accept(ItemVisitor visitor) {
        visitor.visit(this);
    }
}

class InvoiceVisitor implements ItemVisitor {
    @Override
    public void visit(PhysicalProduct physicalProduct) {
        System.out.println("Printing invoice for physical product: " + physicalProduct.name);
    }

    @Override
    public void visit(DigitalProduct digitalProduct) {
        System.out.println("Printing invoice for digital product: " + digitalProduct.name);
    }

    @Override
    public void visit(GiftCard giftCard) {
        System.out.println("Printing invoice for gift card: " + giftCard.code);
    }
}

class ShippingCostVisitor implements ItemVisitor {
    @Override
    public void visit(PhysicalProduct physicalProduct) {
        System.out.println("Shipping cost for physical product: $" + (physicalProduct.weight * 0.5));
    }

    @Override
    public void visit(DigitalProduct digitalProduct) {
        System.out.println("No shipping cost for digital product: " + digitalProduct.name);
    }

    @Override
    public void visit(GiftCard giftCard) {
        System.out.println("No shipping cost for gift card: " + giftCard.code);
    }
}

public class Main {
    public static void main(String[] args) {
        List<Item> items = new ArrayList<>();
        Item physicalProduct = new PhysicalProduct("Laptop", 2.5);
        Item digitalProduct = new DigitalProduct("E-book", 0.1);
        Item giftCard = new GiftCard("GC12345", 50.0);

        items.add(physicalProduct);
        items.add(digitalProduct);
        items.add(giftCard);

        ItemVisitor invoiceVisitor = new InvoiceVisitor();
        ItemVisitor shippingCostVisitor = new ShippingCostVisitor();

        for (Item item : items) {
            item.accept(invoiceVisitor);
            item.accept(shippingCostVisitor);
        }
    }
}

File: 04_BehaviouralDesignPatterns/07_VisitorPattern/cpp/Main.cpp
#include <iostream>
#include <vector>
#include <memory>

// Forward declarations
class PhysicalProduct;
class DigitalProduct;
class GiftCard;

// Visitor interface
class ProductVisitor {
public:
    virtual void visit(const PhysicalProduct& product) = 0;
    virtual void visit(const DigitalProduct& product) = 0;
    virtual void visit(const GiftCard& card) = 0;
    virtual ~ProductVisitor() = default;
};

// Product interface
class Product {
public:
    virtual void accept(ProductVisitor& visitor) = 0;
    virtual ~Product() = default;
};

// Concrete products
class PhysicalProduct : public Product {
public:
    void accept(ProductVisitor& visitor) override {
        visitor.visit(*this);
    }

    double getWeight() const {
        return 2.5; // Example weight in kg
    }

    double getSize() const {
        return 10.0; // Example size in cubic meters
    }
};

class DigitalProduct : public Product {
public:
    void accept(ProductVisitor& visitor) override {
        visitor.visit(*this);
    }

    double getFileSize() const {
        return 1.5; // Example size in GB
    }
};

class GiftCard : public Product {
public:
    void accept(ProductVisitor& visitor) override {
        visitor.visit(*this);
    }

    double getValue() const {
        return 50.0; // Example value in dollars
    }
};

// Concrete visitors
class InvoicePrinter : public ProductVisitor {
public:
    void visit(const PhysicalProduct& product) override {
        std::cout << "Printing invoice for physical product:" << std::endl;
        std::cout << "Weight: " << product.getWeight() << " kg" << std::endl;
        std::cout << "Size: " << product.getSize() << " cubic meters" << std::endl;
    }

    void visit(const DigitalProduct& product) override {
        std::cout << "Printing invoice for digital product:" << std::endl;
        std::cout << "File size: " << product.getFileSize() << " GB" << std::endl;
    }

    void visit(const GiftCard& card) override {
        std::cout << "Printing invoice for gift card:" << std::endl;
        std::cout << "Value: $" << card.getValue() << std::endl;
    }
};

class ShippingCalculator : public ProductVisitor {
public:
    void visit(const PhysicalProduct& product) override {
        double cost = 5.99 + (product.getWeight() * 2.0);
        std::cout << "Shipping cost for physical product: $" << cost << std::endl;
    }

    void visit(const DigitalProduct&) override {
        std::cout << "No shipping cost for digital product" << std::endl;
    }

    void visit(const GiftCard&) override {
        std::cout << "No shipping cost for gift card" << std::endl;
    }
};

class DiscountCalculator : public ProductVisitor {
public:
    void visit(const PhysicalProduct&) override {
        std::cout << "Standard 5% discount for physical product" << std::endl;
    }

    void visit(const DigitalProduct&) override {
        std::cout << "No discount for digital product" << std::endl;
    }

    void visit(const GiftCard& card) override {
        double discount = card.getValue() * 0.1; // 10% discount
        std::cout << "Gift card discount: $" << discount << std::endl;
    }
};

int main() {
    // Create products
    std::vector<std::unique_ptr<Product>> products;
    products.push_back(std::make_unique<PhysicalProduct>());
    products.push_back(std::make_unique<DigitalProduct>());
    products.push_back(std::make_unique<GiftCard>());

    // Create visitors
    InvoicePrinter invoicePrinter;
    ShippingCalculator shippingCalculator;
    DiscountCalculator discountCalculator;

    std::cout << "Processing all products:\n" << std::endl;

    // Process each product with all visitors
    for (const auto& product : products) {
        product->accept(invoicePrinter);
        product->accept(shippingCalculator);
        product->accept(discountCalculator);
        std::cout << "------------------------" << std::endl;
    }

    return 0;
}